src/server/pachyderm_test.go

```diff
@@ -5551,6 +5552,251 @@ func TestJoinInput(t *testing.T) {
 	}
 }
 
+func TestGroupInput(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration tests in short mode")
+	}
+
+	c := tu.GetPachClient(t)
+	require.NoError(t, c.DeleteAll())
+
+	t.Run("Basic", func(t *testing.T) {
+		repo := tu.UniqueString("TestGroupInput")
+		require.NoError(t, c.CreateRepo(repo))
+		numFiles := 16
+		for i := 0; i < numFiles; i++ {
+			_, err := c.PutFile(repo, "master", fmt.Sprintf("file.%4b", i), strings.NewReader(fmt.Sprintf("%d\n", i)))
+			require.NoError(t, err)
+		}
+
+		pipeline := "group-pipeline"
+		require.NoError(t, c.CreatePipeline(
+			pipeline,
+			"",
+			[]string{"bash"},
+			[]string{},
+			&pps.ParallelismSpec{
+				Constant: 1,
+			},
+			client.NewGroupInput(
+				client.NewPFSInputOpts("", repo, "", "/file-?.(?)(?)(?)(?)", "", "$3", false),
+			),
+			"",
+			false,
+		))
+
+		jobs, err := c.FlushJobAll([]*pfs.Commit{client.NewCommit(repo, "master")}, nil)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(jobs))
+
+		// We're grouping by the third digit in the filename
+		// for 0 and 1, this is just a space
+		// then we should see the 8 files with a one there, and the 6 files with a zero there
+		expected := `
+/file-0.   0
+/file-0.   1
+
+/file-0.  10
+/file-0.  11
+/file-0. 110
+/file-0. 111
+/file-0.1010
+/file-0.1011
+/file-0.1110
+/file-0.1111
+
+/file-0. 100
+/file-0. 101
+/file-0.1000
+/file-0.1001
+/file-0.1100
+/file-0.1101
+
+`
```

Comment:
It's a little weird to implement this as a string comparison rather than a more expressive type although I get where it comes from since this was the manual test you were doing initially. I think if you just have 3 `[]string` one for each datum and include the file names in there it'd be a little cleaner. I'm worried about the test potentially breaking when someone modifies it due to silly stuff like newlines.